File: Day1_PointerBasics/hard/hard_03.cpp
/*
 * FILE: hard_03.cpp
 * TOPIC: Pointer Basics - Day 1
 * DIFFICULTY: Hard
 *
 * TASK: Pass a pointer into a function to modify the original variable
 * --------------------------------------------------------------------
 * You may NOT use references (&) in the function signature - only raw pointers.
 *
 * 1. Write a function:
 *      void doubleIt(int* ptr)
 *    that multiplies the value at `ptr` by 2.
 *
 * 2. Write a function:
 *      void addFive(int* ptr)
 *    that adds 5 to the value at `ptr`.
 *
 * 3. In main():
 *      a) Declare int n = 10.
 *      b) Call doubleIt(&n).   Print n.
 *      c) Call addFive(&n).    Print n.
 *      d) Call doubleIt(&n).   Print n.
 *
 * EXPECTED OUTPUT:
 *   After doubleIt: 20
 *   After addFive : 25
 *   After doubleIt: 50
 */

#include <iostream>
using namespace std;

// TODO: Write your function declarations and definitions here.

void doubleIt(int* ptr) {
    *ptr = *ptr * 2;
}

void addFive(int* ptr) {
    *ptr = *ptr + 5;
}

int main() {
    // TODO: Write your solution here.
    int n = 10;
    doubleIt(&n);
    cout << "After doubleIt:" << n << "\n";
    addFive(&n);
    cout << "After addFive :" << n << "\n";
    doubleIt(&n);
    cout << "After doubleIt:" << n << "\n";
    return 0;
}

File: Day1_PointerBasics_Go/hard/hard_03.go
// FILE: hard_03.go
// TOPIC: Pointer Basics - Day 1 (Go)
// DIFFICULTY: Hard
//
// TASK: Pointer aliasing — two pointers to the same memory cell
// -------------------------------------------------------------
// When two pointers hold the same address, a write through either one
// is visible through the other. This is called aliasing.
//
// 1. Declare: x := 42
// 2. Declare: p := &x  and  q := &x   (both point to x)
// 3. Print both *p and *q — they should both be 42.
// 4. Write 100 through p (*p = 100).
//    Print *q — it should also be 100 (same memory!).
// 5. Write 200 through q (*q = 200).
//    Print x directly — it should be 200.
// 6. Print whether p == q (they should hold the same address).
//
// EXPECTED OUTPUT:
//   Initial : *p=42, *q=42
//   After *p=100: *q=100
//   After *q=200: x=200
//   p == q: true

package main

import "fmt"

func main() {
	// TODO: Write your solution here.

}
